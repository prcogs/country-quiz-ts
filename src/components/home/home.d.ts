
type Regions = {
   [key: string]: string
}
