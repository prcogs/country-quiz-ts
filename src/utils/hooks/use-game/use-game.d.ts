
/* eslint-disable no-unused-vars */
type Countries = {
   capital: string[]
   flags: {
      svg?: string,
      png?: string
   }
   name: {
      common: string
   }
   [index: string]: unknown;
}

type CountriesParsed = {
   capital: string
   flags: string | undefined
   name: string
}

type PropsGameProvider = {
   children: ReactNode
}
