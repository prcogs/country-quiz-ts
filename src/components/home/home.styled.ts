import styled from 'styled-components'

import { color } from 'style/color'


const StyledHome = styled.div`
   display: flex;
   flex-direction: column;
   align-items: center;
   justify-content: space-between;
   height: 100%;
`

const Select = styled.select`
   font-size: 18px;
`

const Container = styled.div`
   display: flex;
   justify-content: space-around;
   width: 100%;
`

const Title = styled.h2`
   font-weight: 700;
   font-size: 24px;
   line-height: 36px;
   color: ${color['primary-300']};
`

const Button = styled.button`
   width: 100%;
   font-weight: 700;
   font-size: 24px;
   line-height: 36px;
   align-self: flex-end;

   background-color: ${color['secondary-400']};
   border-radius: 8px;
   padding: 12px 0;
   color: ${color.blank};

   transition: background-color ease-in-out 200ms;

   &:hover {
      background-color: ${color['secondary-500']}
   }

   &.disable {
      cursor: not-allowed;
   }
   `

const Span = styled.span`
   color: ${color.error};
`
export const S = {
   StyledHome,
   Select,
   Container,
   Title,
   Button,
   Span,
}
