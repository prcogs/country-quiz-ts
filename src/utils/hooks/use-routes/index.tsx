import { createContext, ReactNode, useContext, useState } from 'react';

import { routes } from 'utils/constants/routes';


type PropsRoutesProvider = {
   children: ReactNode
}

type Route = {
   route: number,
   changeRoute: () => void
}

const routesContext = createContext<Route>({
   route: routes.HOME,
   changeRoute: () => {},
});

export function useProvideRoutes() {
   const { HOME, GAME, RESULT } = routes
   const [route, setRoute] = useState(HOME)

   const changeRoute = () => {
      switch (route) {
         case HOME:
            setRoute(GAME)
            break;
         case GAME:
            setRoute(RESULT)
            break;
         case RESULT:
            setRoute(HOME)
            break;
         default:
            setRoute(HOME)
            break;
      }
   }

   return {
      route,
      changeRoute,
   };
}

export function ProvideRoutes({ children }: PropsRoutesProvider) {
   const handleRoutes = useProvideRoutes();
   return <routesContext.Provider value={handleRoutes}>{children}</routesContext.Provider>;
}

export function useRoutes() {
   return useContext(routesContext);
}
