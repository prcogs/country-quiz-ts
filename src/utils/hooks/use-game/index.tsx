import { createContext, useContext, ReactNode, useState } from 'react';

import { useRoutes } from 'utils/hooks/use-routes';
import { getCountries } from 'utils/api/request/getCountries';

type Countries = {
   capital: string[]
   flags: {
      svg?: string,
      png?: string
   }
   name: {
      common: string
   }
}

type CountriesParsed = {
   capital: string
   flags: string | undefined
   name: string
}

type PropsGameProvider = {
   children: ReactNode
}

type GameContext= {
   init: (numberQuestion: number, region: string) => Promise<boolean | undefined>
   questions: CountriesParsed[][]
   dataGame: {
      numberQuestions: number
      numberGoodResp: number
      actualQuestion: number
   }
   nextQuestion: () => void
   oneMoreGoodAnswer: () => void
   retryGame: () => void
}


const gameContext = createContext<GameContext>({
   init: async (numberQuestion, region) => false,
   questions: [],
   dataGame: {
      numberQuestions: 0,
      numberGoodResp: 0,
      actualQuestion: 0,
   },
   nextQuestion: () => {},
   oneMoreGoodAnswer: () => {},
   retryGame: () => {},
});

export const createQuestions = (countries: CountriesParsed[], numberQuestions: number) => {
   const nbrCountries = numberQuestions * 4
   const randomNumbers: number[] = []

   for (let i = 0; i < nbrCountries; i += 1) {
      let condition = true
      while (condition) {
         const random = Math.floor(Math.random() * countries.length)
         if (randomNumbers.every(nbr => nbr !== random)) {
            randomNumbers.push(random)
            condition = false
         }
      }
   }

   let allQuestions: CountriesParsed[][] = []
   for (let i = 0; i < randomNumbers.length; i += 4) {
      const singleQuestion = [
         countries[randomNumbers[i]],
         countries[randomNumbers[i + 1]],
         countries[randomNumbers[i + 2]],
         countries[randomNumbers[i + 3]],
      ]
      allQuestions = [...allQuestions, singleQuestion]
   }


   return allQuestions
}


export function useProvideGame() {
   const { changeRoute } = useRoutes()
   const [questions, setQuestions] = useState<CountriesParsed[][]>([])
   const [dataGame, setDataGame] = useState({
      numberQuestions: 0,
      numberGoodResp: 0,
      actualQuestion: 0,
   })

   const init = async (numberQuestions: number, region: string) => {
      const { data, error } = await getCountries(region)


      if (data) {
         const parsedData: CountriesParsed[] = data
            .filter(({ capital, flags, name }: Countries) => capital && flags && name)
            .map(({ capital, flags, name }: Countries) => ({
               capital: capital[0],
               flags: flags?.svg || flags?.png,
               name: name?.common,
            }))

         setQuestions(createQuestions(parsedData, numberQuestions))
         setDataGame({ ...dataGame, numberQuestions })

         return true
      }

      if (error) return false

      return false
   }

   const nextQuestion = () => {
      if (dataGame.actualQuestion + 1 >= dataGame.numberQuestions) changeRoute()
      else setDataGame({ ...dataGame, actualQuestion: dataGame.actualQuestion + 1 })
   }

   const oneMoreGoodAnswer = () => setDataGame({ ...dataGame, numberGoodResp: dataGame.numberGoodResp + 1 })

   const retryGame = () => {
      setDataGame({
         numberQuestions: 0,
         numberGoodResp: 0,
         actualQuestion: 0,
      })

      changeRoute()
   }

   return {
      init,
      questions,
      dataGame,
      nextQuestion,
      oneMoreGoodAnswer,
      retryGame,
   };
}

export function ProvideGame({ children }: PropsGameProvider) {
   const handleGameStage = useProvideGame();
   return <gameContext.Provider value={handleGameStage}>{children}</gameContext.Provider>;
}

export function useGame() {
   return useContext(gameContext);
}
