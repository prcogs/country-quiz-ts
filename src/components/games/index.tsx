import { useEffect, useState } from 'react'

import { useGame } from 'utils/hooks/use-game'

import { S } from 'components/games/games.styled'

const type = {
   CAPITAL: 1,
   FLAG: 2,
}

const random = () => Math.floor(Math.random() * 4)
const letters = ['A', 'B', 'C', 'D']


const Game = ({ data, isPlaying, handleAnswer, wichType }: Game) => {
   const [goodAnswer, setGoodAnswer] = useState(data[random()])
   const [nameOfClick, setidxOfClick] = useState<string | null>(null)

   const handleClick = (name: string) => {
      handleAnswer(name === goodAnswer.name)
      setidxOfClick(name)
   }

   useEffect(() => {
      setGoodAnswer(data[random()])
   }, [data])

   const getClassName = (name: string) => {
      if (!isPlaying) {
         if (name === goodAnswer.name) return 'success'
         if (nameOfClick === name) return 'error'
         return 'none'
      }

      return ''
   }


   return (
      <S.StyledGame>
         {wichType === type.CAPITAL
            ? <S.Question>{`${goodAnswer.capital} is the capital of`}</S.Question>
            : (
               <S.Question>
                  <S.Img src={goodAnswer.flags} alt="Flag" width={100} />
                  Wich country does this flag belong to
               </S.Question>
            )}

         <S.Cards>
            {data.map(({ name }, idx) => (
               <S.Card key={name}>
                  <button
                     type="button"
                     disabled={!isPlaying}
                     onClick={() => handleClick(name)}
                     className={getClassName(name)}
                  >
                     <span>{letters[idx]}</span>

                     {name}

                     {!isPlaying
                        && (
                           <S.SpanImage>
                              {name === goodAnswer.name
                                 ? <img src="/icons/check-solid.svg" alt="Icon check" />
                                 : <img src="/icons/xmark-solid.svg" alt="Icon cross" />}
                           </S.SpanImage>
                        )}
                  </button>
               </S.Card>
            ))}
         </S.Cards>
      </S.StyledGame>
   )
}

export const Games = () => {
   const [isPlaying, setIsPlaying] = useState(true)
   const { questions, dataGame, nextQuestion, oneMoreGoodAnswer } = useGame()

   const handlePlaying = () => {
      setIsPlaying(prevState => !prevState)
   }

   const handleAnswer = (resp: boolean) => {
      if (resp) oneMoreGoodAnswer()
      handlePlaying()
   }

   const handleNext = () => {
      handlePlaying()
      nextQuestion()
   }

   return (
      <S.StyledGames>
         <S.Span>{`${dataGame.actualQuestion + 1} / ${dataGame.numberQuestions}`}</S.Span>

         <Game
            data={questions[dataGame.actualQuestion]}
            isPlaying={isPlaying}
            handleAnswer={handleAnswer}
            wichType={dataGame.actualQuestion % 2 === 0 ? type.CAPITAL : type.FLAG}
         />

         {!isPlaying && <S.Button type="button" onClick={() => handleNext()}>Next</S.Button>}
      </S.StyledGames>
   )
}
