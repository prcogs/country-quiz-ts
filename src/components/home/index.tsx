import { useRef, useState } from 'react';

import { useGame } from 'utils/hooks/use-game';
import { useRoutes } from 'utils/hooks/use-routes'

import { S } from './home.styled'


function getKeyByValue(object: Regions, value: string) {
   return Object.keys(object).find(key => object[key] === value);
}

const regions: Regions = {
   africa: 'Africa',
   americas: 'Americas',
   asia: 'Asia',
   europe: 'Europe',
}

export const Home = () => {
   const { changeRoute } = useRoutes()
   const { init } = useGame()

   const nbRef = useRef<HTMLSelectElement | null>(null)
   const regionRef = useRef<HTMLSelectElement | null>(null)

   const [error, setError] = useState<string | null>(null)
   const [isLoading, setIsLoading] = useState(false)

   const handlePlay = async () => {
      setIsLoading(true)

      const nbr = Number(nbRef?.current?.value) >= 0 && Number(nbRef?.current?.value) <= 10 ? Number(nbRef?.current?.value) : 5
      const region = regionRef?.current?.value || 'europe'
      const go = await init(nbr, getKeyByValue(regions, region) || 'europe')

      setIsLoading(false)

      if (go) changeRoute()
      else setError('Error, please retry')
   }


   return (
      <S.StyledHome data-testid="home">
         <S.Title>Hi, do you want play ?</S.Title>

         <S.Container>
            <S.Select name="nb" id="nb" ref={nbRef}>
               {Array.from({ length: 10 }, (v, i) => i + 1).map(nb => (
                  <option value={nb} key={`${nb}__`}>{nb}</option>
               ))}
            </S.Select>

            <S.Select name="region" id="region" ref={regionRef}>
               {Object.values(regions).map(region => (
                  <option value={region} key={`${region}__`}>{region}</option>
               ))}
            </S.Select>
         </S.Container>

         {error && <S.Span data-testid="error">{error}</S.Span>}

         <S.Button
            type="button"
            className={isLoading ? 'disable' : ''}
            onClick={() => handlePlay()}
            disabled={isLoading}
            data-testid="btn"
         >
            {isLoading ? 'Loading...' : 'Play'}
         </S.Button>
      </S.StyledHome>
   )
}
