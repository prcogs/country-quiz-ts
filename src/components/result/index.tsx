import { useGame } from 'utils/hooks/use-game'

import { S } from 'components/result/result.styled'


export const Result = () => {
   const { dataGame, retryGame } = useGame()

   return (
      <S.StyledResult>
         <S.Title>Results</S.Title>

         <S.Para>
            {'You got '}
            <span>{dataGame.numberGoodResp}</span>
            {` / ${dataGame.numberQuestions} correct answer`}
         </S.Para>

         <S.Button type="button" onClick={() => retryGame()}>Try again</S.Button>
      </S.StyledResult>
   )
}
