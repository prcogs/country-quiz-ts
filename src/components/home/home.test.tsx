import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { useGame } from 'utils/hooks/use-game'
import { Home } from 'components/home'


jest.mock('utils/hooks/use-game', () => ({
   useGame: jest.fn(),
}))

const fct = (boolean: boolean) => () => ({
   init: jest.fn((a: number, b: string | undefined) => Promise.resolve(boolean)),
})

describe('home', () => {
   // beforeAll(() => {
   //    (useGame as jest.Mock)
   //       .mockImplementation(fct(false))
   //       .mockImplementationOnce(fct(true))
   //       .mockImplementationOnce(fct(true))
   // })

   // afterAll(() => {
   //    console.log((useGame as jest.Mock).mock.calls.length, 'ddddddd')
   // })

   test('should render correctly', () => {
      (useGame as jest.Mock).mockImplementation(() => {
         return {
            init: jest.fn(),
         }
      })

      expect(render(<Home />)).toMatchSnapshot()
   });


   test('btn is call & error attempt api data', async () => {
      (useGame as jest.Mock).mockImplementation(fct(false))

      render(<Home />)

      const btn = screen.getByTestId<HTMLButtonElement>('btn')

      await waitFor(() => fireEvent.click(btn))
      expect(btn.disabled).toBe(true)

      await waitFor(() => screen.getByTestId<HTMLSpanElement>('error'))
      expect(btn.disabled).toBe(false)
      expect(screen.getByTestId<HTMLSpanElement>('error')).toBeInTheDocument()
   });


   test('btn is call & start quiz', async () => {
      (useGame as jest.Mock).mockImplementation(fct(true))

      const { getByTestId, queryByTestId } = render(<Home />)

      const btn = getByTestId('btn')

      await waitFor(() => fireEvent.click(btn))
      expect((btn as HTMLButtonElement).disabled).toBe(true)

      expect(queryByTestId('error')).toBe(null)
      await waitFor(() => {
         expect((btn as HTMLButtonElement).disabled).toBe(false)
      });
   });
})
