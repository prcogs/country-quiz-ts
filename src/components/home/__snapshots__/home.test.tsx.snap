// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bczRLJ dgFfCE"
        data-testid="home"
      >
        <h2
          class="sc-hKMtZM ecwSBU"
        >
          Hi, do you want play ?
        </h2>
        <div
          class="sc-dkzDqf TkbfV"
        >
          <select
            class="sc-gsnTZi HINSD"
            id="nb"
            name="nb"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
          </select>
          <select
            class="sc-gsnTZi HINSD"
            id="region"
            name="region"
          >
            <option
              value="Africa"
            >
              Africa
            </option>
            <option
              value="Americas"
            >
              Americas
            </option>
            <option
              value="Asia"
            >
              Asia
            </option>
            <option
              value="Europe"
            >
              Europe
            </option>
          </select>
        </div>
        <button
          class="sc-eCYdqJ cpGAJR"
          data-testid="btn"
          type="button"
        >
          Play
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bczRLJ dgFfCE"
      data-testid="home"
    >
      <h2
        class="sc-hKMtZM ecwSBU"
      >
        Hi, do you want play ?
      </h2>
      <div
        class="sc-dkzDqf TkbfV"
      >
        <select
          class="sc-gsnTZi HINSD"
          id="nb"
          name="nb"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
        <select
          class="sc-gsnTZi HINSD"
          id="region"
          name="region"
        >
          <option
            value="Africa"
          >
            Africa
          </option>
          <option
            value="Americas"
          >
            Americas
          </option>
          <option
            value="Asia"
          >
            Asia
          </option>
          <option
            value="Europe"
          >
            Europe
          </option>
        </select>
      </div>
      <button
        class="sc-eCYdqJ cpGAJR"
        data-testid="btn"
        type="button"
      >
        Play
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
